File: 2022/day06/cpp/solution.cpp
#include <iostream>
#include <fstream>
#include <algorithm>
#include <set>
using namespace std;

int solution(string line, int n) {
    for (int i = 0; i < line.length()-n; i++) {
        set<char> uniq(line.begin()+i, line.begin()+i+n);
        if (uniq.size() == n) {
            return i+n;
        }
    }
    return 0;
}

int main(int argc, char** argv) {
    string infile = argc == 2 ? argv[1] : "input.txt";
    ifstream file(infile);
    string line;
    getline(file, line);
    cout << "Part 1: " << solution(line, 4) << endl;
    cout << "Part 2: " << solution(line, 14) << endl;
    file.close();
    return 0;
}

File: 2022/day06/python/solution.py
import sys
infile = sys.argv[1] if len(sys.argv) > 1 else "../input.txt"

with open(infile) as f:
    line = f.readline().rstrip()

def solution(n):
    for i in range(len(line)-n):
        if len(set(line[i:i+n])) == n:
            return i+n
    return 0

print(f"Part 1: {solution(4)}")
print(f"Part 2: {solution(14)}")
